theseus/geometry/se3.py

```diff
@@ -284,26 +284,88 @@ def vee(matrix: torch.Tensor) -> torch.Tensor:
 
     # only added to avoid casting downstream
     def copy(self, new_name: Optional[str] = None) -> "SE3":
         return cast(SE3, super().copy(new_name=new_name))
 
+    def _transform_shape_check(self, point: Union[Point3, torch.Tensor]):
+        err_msg = "SE3 can only transform 3-D vectors."
+        if isinstance(point, torch.Tensor):
+            if not point.ndim == 2 or point.shape[1] != 3:
+                raise ValueError(err_msg)
+        elif point.dof() != 3:
+            raise ValueError(err_msg)
+        if (
+            point.shape[0] != self.shape[0]
+            and point.shape[0] != 1
+            and self.shape[0] != 1
+        ):
+            raise ValueError(
+                "Input point batch size is not broadcastable with group batch size."
+            )
+
     def transform_to(
         self,
         point: Union[Point3, torch.Tensor],
         jacobians: Optional[List[torch.Tensor]] = None,
     ) -> Point3:
-        raise NotImplementedError
+        self._transform_shape_check(point)
+        batch_size = max(self.shape[0], point.shape[0])
+        if isinstance(point, torch.Tensor):
+            p = point.view(-1, 3, 1)
+        else:
+            p = point.data.view(-1, 3, 1)
+
+        ret = Point3(data=(self[:, :, :3] @ p).view(-1, 3))
+        ret += self[:, :, 3]
+
+        if jacobians is not None:
+            self._check_jacobians_list(jacobians)
+            # Right jacobians for SE(3) are computed
+            Jg = torch.zeros(batch_size, 3, 6, dtype=self.dtype, device=self.device)
+            Jg[:, :, :3] = self[:, :, :3]
+            Jg[:, :, 3:] = -self[:, :, :3] @ SO3.hat(p)
+            # Jacobians for point
+            Jpnt = Jg[:, :, :3]
+
+            jacobians.extend([Jg, Jpnt])
+
+        return ret
 
     def transform_from(
         self,
         point: Union[Point3, torch.Tensor],
         jacobians: Optional[List[torch.Tensor]] = None,
     ) -> Point3:
-        raise NotImplementedError
+        self._transform_shape_check(point)
+        batch_size = max(self.shape[0], point.shape[0])
+        if isinstance(point, torch.Tensor):
+            p = point.view(-1, 3, 1)
+        else:
+            p = point.data.view(-1, 3, 1)
+
+        temp = p - self[:, :, 3:]
+        ret = Point3(data=(self[:, :, :3].transpose(1, 2) @ temp).view(-1, 3))
+
+        if jacobians is not None:
+            self._check_jacobians_list(jacobians)
+            # Right jacobians for SE(3) are computed
+            Jg = torch.zeros(batch_size, 3, 6, dtype=self.dtype, device=self.device)
+            Jg[:, 0, 0] = -1
+            Jg[:, 1, 1] = -1
+            Jg[:, 2, 2] = -1
+            Jg[:, 0, 4] = -ret[:, 2]
+            Jg[:, 1, 3] = ret[:, 2]
+            Jg[:, 0, 5] = ret[:, 1]
+            Jg[:, 2, 3] = -ret[:, 1]
+            Jg[:, 1, 5] = -ret[:, 0]
+            Jg[:, 2, 4] = ret[:, 0]
+            # Jacobians for point
+            Jpnt = self[:, :, :3].transpose(1, 2).expand(batch_size, 3, 3)
```

Comment:
Is this still relevant given you added automatic Jacobians in follow up PRs?

Reply:
Do we have a preference or do we want to let users pick which one they want (since you added support for automatic Jacobians in follow up PRs)?

Reply:
As discussed offline,  for `AutodiffCostFunction` we actually want sparse jacobians for efficiency, but we'll need to look into using `vmap` to replace the current implementation using `torch.autograd.jacobian` (or some other feasible alternative). This is not an issue when doing `loss.backward()`.

Reply:
I would like to keep `LieGroup.project()` as it is since this is a general formation to project `jacobian` and `gradient` from the `Euclidean space` to `Lie group tangent space`. Instead, we can write another function to handle automatic `sparse jacobian` calculation.

Reply:
Please refer to #77.

Reply:
To be clear, I was referring to eventually changing the implementation of `AutodiffCostFunction` to avoid computing dense gradients, but I don't anticipate doing that would require any changes to `project()`.

Reply:
Agreed.